File: DSA/fundamentals/arrays_and_lists/part2/javascript/judge.js
const Solution = require('./solution');

class Judge {
    runTests() {
        const testCasesC = [
            ["Plateau", "Haven", "Insomnia", "Grand Trad", "Plateau", "Sunville"],
            ["Plateau", "Haven", "Insomnia", "Grand Trad", "Sunville", "Bronze City", "Sunville"],
        ];

        let passes = 0;
        for (let i = 0; i < testCasesC.length; i++) {
            if (this.test(testCasesC[i], this.testCaseSolutionOutput(testCasesC[i]))) {
                passes++;
            }
        }
        console.log(`Passed ${passes} of ${testCasesC.length} tests.`);
    }

    test(c, expected) {
        const solution = new Solution();
        const output = solution.dispatchTo(c);
        const passed = output === expected;

        console.log(`c: ${JSON.stringify(c)}`);
        console.log(`Expected Output: ${expected}`);
        console.log(`Your Output: ${output}`);
        console.log(passed ? "Test PASSED" : "Test FAILED");
        console.log("-----------");
        return passed;
    }

    testCaseSolutionOutput(c) {
        const seen = [];
        for (const city of c) {
            if (seen.includes(city)) {
                return city;
            }
            seen.push(city);
        }
        return "";
    }
}

module.exports = Judge;

File: DSA/fundamentals/arrays_and_lists/part2/cpp/Judge.cpp
#include <iostream>
#include <vector>
#include <string>
#include "Solution.cpp"
using namespace std;

class Judge {
public:
    void runTests() {
        vector<vector<string>> testCasesC = {
            {"Plateau", "Haven", "Insomnia", "Grand Trad", "Plateau", "Sunville"},
            {"Plateau", "Haven", "Insomnia", "Grand Trad", "Sunville", "Bronze City", "Sunville"}
        };
        
        runTests(testCasesC);
    }

    void runTests(vector<vector<string>>& testCasesC) {
        int passes = 0;
        for (int i = 0; i < testCasesC.size(); i++) {
            string expected = testCaseSolutionOutput(testCasesC[i]);
            if (test(testCasesC[i], expected)) {
                passes++;
            }
        }
        cout << "Passed " << passes << " of " << testCasesC.size() << " tests." << endl;
    }

private:
    bool test(vector<string>& c, string expected) {
        Solution sol;
        string output = sol.dispatchTo(c);
        bool passed = (output == expected);
        
        cout << "c: [";
        for (int i = 0; i < c.size(); i++) {
            cout << "\"" << c[i] << "\"";
            if (i < c.size() - 1) cout << ", ";
        }
        cout << "]" << endl;
        cout << "Expected Output: " << expected << endl;
        cout << "Your Output: " << output << endl;
        
        cout << (passed ? "Test PASSED" : "Test FAILED") << endl;
        cout << "-----------" << endl;
        return passed;
    }

    string testCaseSolutionOutput(vector<string> c) {
        vector<string> seen;
        for (const string& city : c) {
            for (const string& seenCity : seen) {
                if (seenCity == city) {
                    return city;
                }
            }
            seen.push_back(city);
        }
        return "";
    }
};